website/src/documentation/io/developing-io-overview.md

```diff
@@ -0,0 +1,171 @@
+   for additional style guide recommendations.
+
+
+## Implementation options
+
+### Sources
+
+For bounded (batch) sources, there are currently two options for creating a Beam
+source:
+
+1. Use `ParDo` and `GroupByKey`.  
+
+1. Use the `Source` interface and extend the `BoundedSource` abstract subclass.
+
+`ParDo` is the recommended option, as implementing a `Source` can be tricky. See
+[When to use the Source interface](#when-to-use-source) for a list of some use
+cases where you might want to use a `Source` (such as
+[dynamic work rebalancing]({{ site.baseurl }}/blog/2016/05/18/splitAtFraction-method.html)).
+
+(Java only) For unbounded (streaming) sources, you must use the `Source`
+interface and extend the `UnboundedSource` abstract subclass. `UnboundedSource`
+supports features that are useful for streaming pipelines, such as
+checkpointing.
+
+Splittable DoFn is a new sources framework that is under development and will
+replace the other options for developing bounded and unbounded sources. For more
+information, see the
+[roadmap for multi-SDK connector efforts]({{ site.baseurl }}/roadmap/connectors-multi-sdk/).
+
+### When to use the Source interface {#when-to-use-source}
+
+If you are not sure whether to use `Source`, feel free to email the [Beam dev
+mailing list]({{ site.baseurl }}/get-started/support) and we can discuss the
+specific pros and cons of your case.
+
+In some cases, implementing a `Source` might be necessary or result in better
+performance:
+
+* **Unbounded sources:** `ParDo` does not work for reading from unbounded
+  sources.  `ParDo` does not support checkpointing or mechanisms like de-duping
+  that are useful for streaming data sources.  
+
+* **Progress and size estimation:** `ParDo` can't provide hints to runners about
+  progress or the size of data they are reading. Without size estimation of the
+  data or progress on your read, the runner doesn't have any way to guess how
+  large your read will be, and thus if it attempts to dynamically allocate
+  workers, it does not have any clues as to how many workers you may need for
+  your pipeline.  
+
+* **Dynamic work rebalancing:** `ParDo` does not support dynamic work
+  rebalancing, which is used by some readers to improve the processing speed of
+  jobs. Depending on your data source, dynamic work rebalancing might not be
+  possible.  
+
+* **Splitting into parts of particular size recommended by the runner:** `ParDo`
+  does not receive `desired_bundle_size` as a hint from runners when performing
+  initial splitting.
+
+For example, if you'd like to read from a new file format that contains many
+records per file, or if you'd like to read from a key-value store that supports
+read operations in sorted key order.
+
+
+### Sinks
+
+To create a Beam sink, we recommend that you use a single `ParDo` that writes the
+received records to the data store. However, for file-based sinks, you can use
```

Comment:
To create a Beam sink, we recommend that you use a ParDo that writes the received records to the data store and use ParDo, GroupByKey, and other transforms that available in Beam to develop more complex sinks (for example, to support data de-duplication when failures are retried by a runner). For file-based sinks, you can use the `FileBasedSink` abstraction offered by both Java and Python SDKs. Please see language specific guidelines for more details.